core/unit_test/TestExecutionSpace.hpp

```diff
@@ -44,4 +44,57 @@ TEST(TEST_CATEGORY, execution_space_as_class_data_member) {
 }
 #endif
 
+constexpr bool test_execspace_explicit_construction() {
+#ifdef KOKKOS_ENABLE_DEPRECATED_CODE_4
+#ifdef KOKKOS_ENABLE_SERIAL
+  static_assert(std::is_convertible_v<Kokkos::NewInstance, Kokkos::Serial>);
+#endif
+#ifdef KOKKOS_ENABLE_OPENMP
+  static_assert(std::is_convertible_v<int, Kokkos::OpenMP>);
+#endif
+#ifdef KOKKOS_ENABLE_CUDA
+  static_assert(std::is_convertible_v<cudaStream_t, Kokkos::Cuda>);
+#endif
+#ifdef KOKKOS_ENABLE_HIP
+  static_assert(std::is_convertible_v<hipStream_t, Kokkos::HIP>);
+#endif
+#ifdef KOKKOS_ENABLE_HPX
+  static_assert(std::is_convertible_v<Kokkos::Experimental::HPX::instance_mode,
+                                      Kokkos::Experimental::HPX>);
+  static_assert(
+      std::is_convertible_v<hpx::execution::experimental::unique_any_sender<>&&,
+                            Kokkos::Experimental::HPX>);
+#endif
+#else
+#ifdef KOKKOS_ENABLE_SERIAL
+  static_assert(!std::is_convertible_v<Kokkos::NewInstance, Kokkos::Serial>);
+#endif
+#ifdef KOKKOS_ENABLE_OPENMP
+  static_assert(!std::is_convertible_v<int, Kokkos::OpenMP>);
+#endif
+#ifdef KOKKOS_ENABLE_CUDA
+  static_assert(!std::is_convertible_v<cudaStream_t, Kokkos::Cuda>);
+#endif
+#ifdef KOKKOS_ENABLE_HIP
+  static_assert(!std::is_convertible_v<hipStream_t, Kokkos::HIP>);
+#endif
+#ifdef KOKKOS_ENABLE_HPX
+  static_assert(!std::is_convertible_v<Kokkos::Experimental::HPX::instance_mode,
+                                       Kokkos::Experimental::HPX>);
+  static_assert(!std::is_convertible_v<
+                hpx::execution::experimental::unique_any_sender<>&&,
+                Kokkos::Experimental::HPX>);
+#endif
+#endif
+
+#ifdef KOKKOS_ENABLE_SYCL
+  static_assert(
+      !std::is_convertible_v<sycl::queue, Kokkos::Experimental::SYCL>);
+#endif
```

Comment:
I am not a big fan of these backend-specific ifdefs but I don't have a better idea at the moment.
Note that OpenACC also had properly used explicit on a constructor from `int`:
https://github.com/kokkos/kokkos/blob/7d0e3e8184779c6f6b0d3ad7eac0341dbf273a2c/core/src/OpenACC/Kokkos_OpenACC.hpp#L76

Reply:
I agree. Another option was individual tests in the backend specific folders, but that seemed to add a lot more lines.

Reply:
I'll add an OpenACC test to the group.